performance/class-media-library-caching.php

```diff
@@ -0,0 +1,164 @@
+<?php
+/**
+ * Media library caching.
+ *
+ * @package vip-performance
+ */
+
+// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery
+
+namespace Automattic\VIP\Performance;
+
+/**
+ * Media Library Caching class.
+ */
+class Media_Library_Caching {
+	public const MINIMUM_WORDPRESS_VERSION          = '6.4';
+	public const CACHE_GROUP                        = 'mime_types';
+	public const AVAILABLE_MIME_TYPES_CACHE_KEY     = 'vip_available_mime_types';
+	public const USING_DEFAULT_MIME_TYPES_CACHE_KEY = 'vip_using_default_mime_types';
+	public const MAX_POSTS_TO_QUERY_DEFAULT         = 100000;
+
+	/**
+	 * Class initialization.
+	 */
+	public static function init() {
+		global $wp_version;
+
+		/**
+		 * Filters whether the MIME type caching is enabled.
+		 *
+		 * @param bool $cache_mime_types Whether the MIME type caching is enabled. Default false.
+		 */
+		$vip_cache_mime_types = apply_filters( 'vip_cache_mime_types', true );
+
+		if ( $vip_cache_mime_types &&
+			isset( $wp_version ) &&
+			version_compare( $wp_version, self::MINIMUM_WORDPRESS_VERSION, '>=' ) ) {
+			self::enable_post_mime_types_caching();
+		}
+	}
+
+	/**
+	 * Enable MIME type caching.
+	 */
+	private static function enable_post_mime_types_caching() {
+		add_filter( 'pre_get_available_post_mime_types', array( __CLASS__, 'get_cached_post_mime_types' ), 10, 2 );
+		add_action( 'add_attachment', array( __CLASS__, 'update_post_mime_types_cache_on_add' ), 10, 1 );
+		add_action( 'attachment_updated', array( __CLASS__, 'update_post_mime_types_cache_on_edit' ), 10, 3 );
+		add_action( 'delete_attachment', array( __CLASS__, 'update_post_mime_types_cache_on_delete' ), 10 );
+	}
+
+	/**
+	 * Get cached results for get_available_post_mime_types() to avoid a query on every page load.
+	 *
+	 * @param string[]|null $filtered_mime_types An array of MIME types. Default null.
+	 * @param string        $type                The post type name.
+	 * @return array An array of MIME types.
+	 */
+	public static function get_cached_post_mime_types( $filtered_mime_types, $type ) {
+		global $wpdb;
```

Comment:
Minor nitpick - let's move this down since we don't need to call the global yet.